kong/db/schema/others/declarative_config.lua

```diff
@@ -0,0 +1,438 @@
+local utils = require("kong.tools.utils")
+local Errors = require("kong.db.errors")
+local Entity = require("kong.db.schema.entity")
+local Schema = require("kong.db.schema")
+local plugin_loader = require("kong.db.schema.plugin_loader")
+
+
+local null = ngx.null
+
+
+local DeclarativeConfig = {}
+
+
+local core_entities = {
+  "services",
+  "routes",
+  "consumers",
+  "plugins",
+  "upstreams",
+  "targets",
+  "certificates",
+  "snis",
+  "cluster_ca",
+}
+
+
+local all_schemas
+
+
+-- Maps a foreign fields to foreign entity names
+-- e.g. `foreign_references["routes"]["service"] = "services"`
+local foreign_references = {}
+
+-- Maps an entity to entities that foreign-reference it
+-- e.g. `foreign_children["services"]["routes"] = "service"`
+local foreign_children = {}
+
+
+function DeclarativeConfig.pk_string(schema, object)
+  if #schema.primary_key == 1 then
+    return tostring(object[schema.primary_key[1]])
+  else
+    local out = {}
+    for _, k in ipairs(schema.primary_key) do
+      table.insert(out, tostring(object[k]))
+    end
+    return table.concat(out, ":")
+  end
+end
+
+
+--[[
```

Comment:
Do we want to leave this in?

Reply:
it was pretty useful for debugging and the code in this module is tricky. I think the next person who needs to make changes to this module would appreciate having that around. In very particular cases we have kept commented-out code around, I thought this was one of them. But I can remove it if truly necessary.